lib/gst/player/gstplayer.c

```diff
@@ -1121,6 +1167,464 @@ element_cb (GstBus * bus, GstMessage * msg, gpointer user_data)
+}
+
+static void
+gst_player_stream_info_update (GstPlayer *self, GstPlayerStreamInfo *s)
+{
+  if (GST_IS_PLAYER_VIDEO_INFO(s))
+    gst_player_video_info_update (self, s);
+
+  if (GST_IS_PLAYER_AUDIO_INFO(s))
+    gst_player_audio_info_update (self, s);
+
+  if (GST_IS_PLAYER_SUBTITLE_INFO(s))
+    gst_player_subtitle_info_update (self, s);
+}
+
+static void
+gst_player_stream_info_update_tags (GstPlayer *self, GstPlayerStreamInfo *s)
+{
+  GstTagList *tags;
+  gint stream_index;
+
+  stream_index = gst_player_stream_info_get_stream_index (s);
+
+  if (GST_IS_PLAYER_VIDEO_INFO(s))
+    g_signal_emit_by_name (self->priv->playbin, "get-video-tags",
+        stream_index, &tags);
+
+  if (GST_IS_PLAYER_AUDIO_INFO(s))
+    g_signal_emit_by_name (self->priv->playbin, "get-audio-tags",
+        stream_index, &tags);
+
+  if (GST_IS_PLAYER_SUBTITLE_INFO(s))
+    g_signal_emit_by_name (self->priv->playbin, "get-text-tags",
+        stream_index, &tags);
+
+  if (s->tags)
+    gst_tag_list_unref (s->tags);
+  s->tags = tags;
+
+  if (s->caps)
+    gst_caps_unref (s->caps);
+  s->caps = get_caps (self, stream_index, G_OBJECT_TYPE(s));
+
+  GST_DEBUG_OBJECT (self, "%s index: %d tags: %p caps: %p",
+      gst_player_stream_info_get_stream_type_nick(s), stream_index,
+      s->tags, s->caps);
+
+  gst_player_stream_info_update (self, s);
+}
+
+static void
+gst_player_streams_info_create (GstPlayer *self, GstPlayerMediaInfo *media_info,
+  const gchar *prop, GType type)
+{
+  gint i;
+  gint total = -1;
+  GstPlayerStreamInfo *s;
+
+  if (!media_info)
+    return;
+
+  g_object_get (G_OBJECT (self->priv->playbin), prop, &total, NULL);
+
+  GST_DEBUG_OBJECT (self, "%s: %d", prop, total);
+
+  for (i = 0; i < total; i++) {
+    s = gst_player_stream_info_find (self, media_info, type, i);
+    if (!s) {
+      s = gst_player_stream_info_new (i, type);
+      media_info->stream_list = g_list_append (media_info->stream_list, s);
+      GST_DEBUG_OBJECT (self, "create %s stream stream_index: %d",
+        gst_player_stream_info_get_stream_type_nick(s), i);
+    }
+    gst_player_stream_info_update_tags (self, s);
+  }
+}
+
+static void
+video_changed_cb (GObject *object, gpointer user_data)
+{
+  GstPlayer *self = GST_PLAYER(user_data);
+
+  g_mutex_lock (&self->priv->lock);
+  gst_player_streams_info_create (self, self->priv->media_info,
+      "n-video", GST_TYPE_PLAYER_VIDEO_INFO);
+  g_mutex_unlock (&self->priv->lock);
+}
+
+static void
+audio_changed_cb (GObject *object, gpointer user_data)
+{
+  GstPlayer *self = GST_PLAYER(user_data);
+
+  g_mutex_lock (&self->priv->lock);
+  gst_player_streams_info_create (self, self->priv->media_info,
+      "n-audio", GST_TYPE_PLAYER_AUDIO_INFO);
+  g_mutex_unlock (&self->priv->lock);
+}
+
+static void
+subtitle_changed_cb (GObject *object, gpointer user_data)
+{
+  GstPlayer *self = GST_PLAYER(user_data);
+
+  g_mutex_lock (&self->priv->lock);
+  gst_player_streams_info_create (self, self->priv->media_info,
```

Comment:
Here (and in the other callbacks) you're modifying existing stream infos but we said the mediainfo/streaminfos are immutable